src/storage/src/source/reclock.rs

```diff
@@ -8,413 +8,668 @@
+
+        // Reclock offsets 1 and 2 to timestamp 1000
+        batch.insert(
+            PART_ID,
+            vec![(1, MzOffset::from(1)), (2, MzOffset::from(2))],
+        );
+        let reclocked_msgs = operator
+            .reclock(&mut batch)
+            .await
+            .flat_map(|(_, msgs)| msgs)
+            .collect_vec();
+        assert_eq!(reclocked_msgs, &[(1, 1000), (2, 1000)]);
+        assert!(batch[&PART_ID].is_empty());
+
+        // Reclock offsets 3 and 4 to timestamp 2000
+        batch.insert(
+            PART_ID,
+            vec![(3, MzOffset::from(3)), (4, MzOffset::from(4))],
+        );
+        let reclocked_msgs = operator
+            .reclock(&mut batch)
+            .await
+            .flat_map(|(_, msgs)| msgs)
+            .collect_vec();
+        assert_eq!(reclocked_msgs, &[(3, 2000), (4, 2000)]);
+        assert!(batch[&PART_ID].is_empty());
+
+        // Compact enough so that we can correctly timestamp only offsets > 2
+        operator.compact(Antichain::from_elem(1000)).await;
+
+        // Reclock offsets 3 and 4 again to see we haven't lost the ability
+        batch.insert(
+            PART_ID,
+            vec![(3, MzOffset::from(3)), (4, MzOffset::from(4))],
+        );
+        let reclocked_msgs = operator
+            .reclock(&mut batch)
+            .await
+            .flat_map(|(_, msgs)| msgs)
+            .collect_vec();
+        assert_eq!(reclocked_msgs, &[(3, 2000), (4, 2000)]);
+        assert!(batch[&PART_ID].is_empty());
+
+        // Attempt to reclock offset 2 which should cause a panic
+        batch.insert(PART_ID, vec![(2, MzOffset::from(2))]);
+        operator
+            .reclock(&mut batch)
+            .await
+            .flat_map(|(_, msgs)| msgs)
+            .collect_vec();
+    }
+
+    #[tokio::test(start_paused = true)]
+    async fn test_concurrency() {
+        const PART_ID: PartitionId = PartitionId::None;
+
+        // Create two operators pointing to the same shard
+        let shared_shard = ShardId::new();
+        let location = PersistLocation {
+            blob_uri: "mem://".to_owned(),
+            consensus_uri: "mem://".to_owned(),
+        };
+        let client = location.open().await.unwrap();
+        let mut op_a = test_operator_with_client(client.clone(), shared_shard).await;
+        let mut op_b = test_operator_with_client(client, shared_shard).await;
+
+        // Reclock a batch from one of the operators
+        let mut batch = HashMap::new();
+
+        // Reclock offsets 1 and 2 to timestamp 1000 from operator A
+        batch.insert(
+            PART_ID,
+            vec![(1, MzOffset::from(1)), (2, MzOffset::from(2))],
+        );
+        let reclocked_msgs = op_a
+            .reclock(&mut batch)
+            .await
+            .flat_map(|(_, msgs)| msgs)
+            .collect_vec();
+        assert_eq!(reclocked_msgs, &[(1, 1000), (2, 1000)]);
+        assert!(batch[&PART_ID].is_empty());
+
```

Comment:
please also `advance` `op_a` here to show that the `op_b` code below still works

Reply:
Done!